Content.Server/Fluids/EntitySystems/FluidSpreaderSystem.cs

```diff
@@ -0,0 +1,193 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using Content.Server.Chemistry.EntitySystems;
+using Content.Server.Fluids.Components;
+using Content.Shared.Chemistry.Components;
+using Content.Shared.Directions;
+using Content.Shared.FixedPoint;
+using Content.Shared.Physics;
+using JetBrains.Annotations;
+using Robust.Shared.Map;
+using Robust.Shared.Physics;
+
+namespace Content.Server.Fluids.EntitySystems;
+
+[UsedImplicitly]
+public sealed class FluidSpreaderSystem : EntitySystem
+{
+    [Dependency] private readonly IMapManager _mapManager = default!;
+    [Dependency] private readonly PuddleSystem _puddleSystem = default!;
+    [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+
+
+    private float _accumulatedTimeFrame;
+    private HashSet<EntityUid> _fluidSpread = new();
+
+    public override void Initialize()
+    {
+        SubscribeLocalEvent<FluidSpreaderComponent, ComponentAdd>((uid, component, _) =>
+            FluidSpreaderAdd(uid, component));
+    }
+
+    public void AddOverflowingPuddle(PuddleComponent puddleComponent, Solution? solution = null)
+    {
+        var puddleSolution = solution;
+        if (puddleSolution == null && !_solutionContainerSystem.TryGetSolution(puddleComponent.Owner,
+                puddleComponent.SolutionName,
+                out puddleSolution)) return;
+
+        if (puddleSolution.CurrentVolume <= puddleComponent.OverflowVolume)
+            return;
+
+        var spreaderComponent = EntityManager.EnsureComponent<FluidSpreaderComponent>(puddleComponent.Owner);
+        spreaderComponent.OverflownSolution = puddleSolution;
+        spreaderComponent.Enabled = true;
+        FluidSpreaderAdd(spreaderComponent.Owner, spreaderComponent);
+    }
+
+    private void FluidSpreaderAdd(EntityUid uid, FluidSpreaderComponent component)
+    {
+        if (component.Enabled)
+            _fluidSpread.Add(uid);
+    }
+
+    public override void Update(float frameTime)
+    {
+        _accumulatedTimeFrame += frameTime;
+
+        if (!(_accumulatedTimeFrame >= 1.0f))
+            return;
+
+        _accumulatedTimeFrame -= 1.0f;
+
+        base.Update(frameTime);
+
+        var spreaders = new HashSet<EntityUid>(_fluidSpread);
+        _fluidSpread = new HashSet<EntityUid>();
+        foreach (var uid in spreaders)
+        {
+            if (EntityManager.Deleted(uid))
+                continue;
```

Comment:
1. You can use a remie queue (RemQueue<T>) and then just add deleted stuff to that. Then afterwards you just iterate over the remqueue. This saves having to allocate the hashset every time.
2. Should also just continue if the spreader is paused, something like this:

```suggestion
        var spreaders = new HashSet<EntityUid>(_fluidSpread);
        _fluidSpread = new HashSet<EntityUid>();
        foreach (var uid in spreaders)
        {
            MetaDataComponent? meta = null;
            
            if (Deleted(uid, meta))
            // RemQueue add here
                continue;
            
            if (Paused(uid, meta))
                continue;
```

The MetaDataComponent allows you to avoid resolving it twice (once for pause and once for deleted).